### DISCUSSION TITLE
Ceiling Fan Speed Control

### DISCUSSION THREAD

**Discussion by BluThunder2k** [Blackymas/NSPanel_HA_Blueprint#791] [Ideas]
I am using NSPanels's to control fan speed and dimming on the light.

Holding in the left button for the light, I get the color/temp/brightness options for the lamp.

Would it be possible to add the similar functionality for fans for when I would hold I the right (fan) button?

HA now shows fans lovelace cards with "high/medium/low/off" buttons, so something similar to that.   

I've created buttons and automations to give me low/medium/high speed buttons on the left swipe page now, but this would be a cleaner look.

Thanks!

**Comment by edwardtfn** [comment]
Looks similar to:
- #404 
- https://github.com/Blackymas/NSPanel_HA_Blueprint/discussions/718#discussioncomment-5786106

**Comment by edwardtfn** [comment]
I've just submitted #897 which will add a fan speed control page. Hopefully this gets all the way to the next release to `main`.

**Reply by Doughboy68** [reply]
Thanks for this. If I want to test a panel, is the dev build based on the main or is it a totally separate which may not contain some of main features ? Sorry if question is dumb but I am not a developer and familiar much with how GitHub works - I notice that the dev build says it is 39 commits behind so just wondering what this means.

**Reply by edwardtfn** [reply]
`dev` is the playground for developers.
You will find there a kind of a preview of the coming version, which could have some new features, but it can happen (rarely) to remove some functionality.
The thing you should be aware when using `dev` is that some functionalities might be incomplete or broken. We will be more than happy if you let us know about some issues you find in `dev`, but in any way we expect to have that environment free of issues.
And be aware that it is quite common that we develop some feature, let's say, with the EU version, but leave to the last minute before the release to replicate that change to the US version, so some things could be broken. We always try to avoid that, but we don't stress if that happens.

> I notice that the dev build says it is 39 commits behind so just wondering what this means.

This is most likely due to some hot fix applied directly to `main` in the past. Currently `dev` have all the functionality you will find in `main` with the addition of new functionality.


This time I'm trying to keep a preview of the next release notes, so maybe you can use that to have an idea about what you will find in `dev`: https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/dev/ReleaseNotes.md
And I can anticipate the new alarm panel is not available yet.

**Reply by Doughboy68** [reply]
Ok, I setup a panel using dev and all seems to work fine. I like the fact that light entities do not show RGB etc. if they are just dimmable. An idea is there could be a few presets attached/linked to a light entity which would then display on that screen, and could be used to call entities so that could control light settings. Eg... I currently use the hard button to call a script which toggles my Family Room lights to a default color... this way if I previously changed them to some weird color, next time my wife comes in the room and turns on they just set to normal. Having a button on that screen I could set to different settings and default color would be nice.

Re: the fan control, there is at the top left a fan down button, fan button and an fan up button, then on very right a fan button.
I suggest swapping the buttons so the up is on the right side instead of the left side. The middle fan button seems to set the fan to 25% (low speed) when pressed and I am not sure I understand why it is there, unless to turn fan on again from that screen (but can press up button)? On my ceiling fan, the up/down buttons cycle through the low,med,high correctly (again, swapping them would make more sense to me). The right button shuts off. Is it possible to make the buttons do off, low, medium, and high? Or is this difficult to do with possible variations of fans? The up/down works fine though if that is easiest way.

Also, can't seem to get into settings screen, from what I can read from the release notes it may have been removed unless I misread? Not a big deal since can be done from Home Assistant.

**Reply by edwardtfn** [reply]
> An idea is there could be a few presets attached/linked to a light entity which would then display on that screen, and could be used to call entities so that could control light settings. Eg... I currently use the hard button to call a script which toggles my Family Room lights to a default color... this way if I previously changed them to some weird color, next time my wife comes in the room and turns on they just set to normal. Having a button on that screen I could set to different settings and default color would be nice.

Could you please [create a new feature request](https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/new) for this specific case?

**Reply by edwardtfn** [reply]
> The middle fan button seems to set the fan to 25% (low speed) when pressed and I am not sure I understand why it is there, unless to turn fan on again from that screen (but can press up button)?

The intention was to have a button to turn on the fan on it's mid point (50%). I will take a look on what is wrong there.

**Reply by edwardtfn** [reply]
> Is it possible to make the buttons do off, low, medium, and high? Or is this difficult to do with possible variations of fans?

That is a bit challenging, as different fans will have different number of speeds. Even set those speed steps is now deprecated in Home Assistant and the recommendation is to always use the percentage, so it covers all the different models.

**Reply by edwardtfn** [reply]
> Also, can't seem to get into settings screen, from what I can read from the release notes it may have been removed unless I misread?

Settings screen is reachable with a long press on the time string on the Home page.
I've changed the engine behind this, moving it from the ESPHome to the Nextion display itself, so if you haven't updated your TFT in the last day or so you might have to do it.

**Reply by Doughboy68** [reply]
Re: fan speed.. I checked again and seems the middle button sets it at 33% which is low speed, my fan actually adjusts reporting to 25% after any setting within this range 0-33. FYI, Medium speed on my Fan seems to adjust back to 74%. I think what is happening is the 33% setting you are doing is being recognized as Low. A percent of 66% would fix this.

I have updated tft today and I cannot long press on time to get the settings. I am using us panel so maybe something not yet implemented in it yet?

**Reply by edwardtfn** [reply]
> I have updated tft today and I cannot long press on time to get the settings. I am using us panel so maybe something not yet implemented in it yet?

That's right. US tft was missing one line of code.
Already fixed in `dev`.

**Reply by Doughboy68** [reply]
Just updated tft. Settings screen working when long press time on US panel.
Fan speed - middle button is working as intended and sets fan 66% (medium). All other buttons work as expected - up/down next speed level, and off.

**Reply by Doughboy68** [reply]
Is it still possible for the US Panel to have the up speed button on the right and the down speed on the left (currently seems backwards)? Seems to make more sense to me that the right side should be up and left is down for button controls.

**Reply by edwardtfn** [reply]
I couldn't find time to look at this, but will give a try when possible.

**Reply by Doughboy68** [reply]
Seen there is an update you did for this, I will try it later when I am home.

**Reply by Doughboy68** [reply]
@edwardtfn I tried the latest dev version with the redistribute icons change, and it looks good but.. Issue is that whenever I turn fan off in that screen, whether it is the off button, arrow down/slower button, or sliding to 0% it actually puts the fan back at 74%. If I exit the screen, the button icon on button screen shows 74% and the fan actually is at medium speed. When I got back in with long press it is 74%. I can speed it up and slow it down, but never can shut off.  I can at least still shut it off on the buttons screen by pressing the button though. It was working in previous version fine before icons were moved, so maybe you inadvertently changed something else that affected this. To reiterate simpler, the fan screen makes it look like it shuts fan off, but it is actually turned back on at 74% (default on level).

**Reply by edwardtfn** [reply]
I will take a look.
And yes, I've changed the engine calling the service as new ESPHome imposes some restrictions for newly added devices to call services directly, forcing me to rebuild all the service calls (including for covers, lights and climate, so I would appreciate if you can also test those).

**Reply by Doughboy68** [reply]
I tested lights dimmer on my dev panel and works... I can slide the slider down to 0% and shuts the light off. The only cover device I have is garage door... will try that later when I get home since didn't have time this morning to test properly. I can try the previous TFT since I made a backup prior, and see if that still works with fan, that way know where issue is... presuming the version I have still works with any other changes you may have done.

**Reply by Doughboy68** [reply]
I tried my garage door which is type cover and with the current dev version it does not bring up a cover screen like the current release version does. Maybe it doesn't bother since there are no positions in this garage door because it is just open/close ? I may not be testing a good example here.

I loaded previous us tft to check fan and the panel does not work at all  so I reverted back to current us tft. I suspect that because of the significant changes made this is why, so not concerned about this. But the fan still does not shut off from the fan screen.

**Reply by edwardtfn** [reply]
I think i know where the issue is. Before i was using the `fan.set_percentage` to set it to 0%, but now I’m using the `fan.turn_off` instead, which should work as this is how the single press stops the fan.

**Reply by edwardtfn** [reply]
> I tried my garage door which is type cover and with the current dev version it does not bring up a cover screen like the current release version does. Maybe it doesn't bother since there are no positions in this garage door because it is just open/close ? I may not be testing a good example here.

That is another feature we implemented recently. The detailed pages are shown only when supported by the entity and will only show the supported features.

**Reply by Doughboy68** [reply]
Ok I will keep an eye out for any changes you make and try.

That is strange how it doesn't stop the fan, according to what you said.

**Reply by edwardtfn** [reply]
I have to look if the service call is correct. I’m in a trip these days, so I don’t have a panel with me and my time is a bit limited, but will invest more time on this next week when back home.

**Reply by Doughboy68** [reply]
Did you get a chance to look at this?

**Reply by Doughboy68** [reply]
I think I figured out the issue. I am not really a coder but I spent quite a while looking at this trying to figure it out. I notice that you move more things directly to the Panel calling services directly via ESPHome and took me a while to track down.

Anyways the issue is in nspanel_esphome.yaml around line 1020, is ending up calling service fan: turn_on with a value of 0%, which in my case using developer tools in home assistant turns my fan on if enter a 0%, which is default 74% (medium speed), which you then will calculate the percent bar down to I think a different percentage (this is irrelevant though). I am not sure if all fans operate this way, but mine does. If you change that line to fan: set_percentage it should fix it, or maybe add some more code/logic in the timer0 in tft to send a key value of stop if speed=0 like you have coded in the esphome.yaml which blanks those settings for fan speed. I will test tomorrow to confirm, but it is late here in western Canada, and wife is sleeping up in bedroom where fan is LOL. 

Thought maybe there is code in the EU nspanel that didn't make it into the US Panel... I checked and the button off in EU contains the proper "domain" for direct control vs US which is an older "page" for blueprint control. But it still calls timer0 with touch release event, so will probably end up turning it on to 0 percent anyways, so that will need to be removed.

fansetting.txt="{\"domain\": \"fan\", \"key\": \"stop\", \"value\": "+va0.txt+", \"entity\": \""+home.entity.txt+"\"}"

US Panel: fansetting.txt="{\"page\": \"fan\", \"component\": \"percentage\", \"value\": "+va0.txt+", \"entity\": \""+home.entity.txt+"\"}"

I still think the timer0 will not work though because it will still end making it turn_on with 0 percentage... maybe an if statement to handle that case. I will test more. 

EDIT Just tested with changes in tft and removing release event timer0 call from off button. Fan turns off with button now. Will look at other down button later.

**Reply by edwardtfn** [reply]
Yeah, I can confirm it is sending a `fan.turn_on` with `percentage: 0` right after the `fan.turn_off`.

**Reply by edwardtfn** [reply]
Thanks a lot for your detailed troubleshooting. You got directly to the point. ;)
I've disabled the timer for that button and it looks to be working fine now.
Please update your TFT and let me know if it is working fine from your side.

**Reply by Doughboy68** [reply]
I cannot test now because I am at work but I did download and look... the nspanel_us.HMI still has the old page message in the button_off and I believe it should be changed to domain (I changed this in my test before and works). Since you are using EU panel you probably missed this. Unless there is old code in the blueprint which handles and this still works, but I think it is intent is to use the domain method.

However, the timer0 seems to contain the main code called by everything on that page, which sends the domain message to control fan, so I would think the message in button_off is redundant and could make changes to timer0 so it checks the speed and sends correct key?

For timer0 I will try this later after work... in timer0 check if fan speed is 0 or not and send appropriate key stop or percentage...  

timer0.en=0
sys0=100*fanslider.val/fanslider.maxval
covx sys0,va0.txt,0,0
fan_value.txt=va0.txt+"%"
if(fanslider.val>fanslider.minval)
{
  fansetting.txt="{\"domain\": \"fan\", \"key\": \"percentage\", \"value\": "+va0.txt+", \"entity\": \""+home.entity.txt+"\"}"
}else
{
  fansetting.txt="{\"domain\": \"fan\", \"key\": \"stop\", \"value\": "+va0.txt+", \"entity\": \""+home.entity.txt+"\"}"
}
printh 92
prints "localevent",0
printh 00
prints fansetting.txt,0
printh 00
printh FF FF FF

And button_off remove the beginning code which sends the domain message, and timer0.en=1 back in touch release event


fanslider.val=fanslider.minval
if(fanslider.val>fanslider.minval)
{
  button_off.pco=59164
  button_down.pco=59164
}else
{
  button_off.pco=10597
  button_down.pco=10597
}
if(fanslider.val<fanslider.maxval)
{
  button_up.pco=59164
}else
{
  button_up.pco=10597
}

I will also test climate later when I load this onto my test panel.

**Reply by edwardtfn** [reply]
On mu tests it isn't calling the second service anymore, so I believe the problem is solved, but I'd like your confirmation before I take this as a closed issue.
Please let me know when you have a chance to test it.

**Reply by Doughboy68** [reply]
It looks like what you did is the same as what I did testing last night - disabling the timer0. But the other button for down, and moving the slider to zero, still calls that timer0 on keypress up. I flashed the code above remotely from work, and will try when I get home soon.

I will also reflash and try the current US panel on dev. But before I do did you test other ways where you can turn fan off... sliding down to zero, and pressing the button down until it is supposed to shut off?

**Reply by Doughboy68** [reply]
I tested and yours doesn't work.  in the button_off, you are sending via fansettings.txt the old page method instead of new domain and it does not appear recognize that method anymore. The EU TFT has correct domain so this is probably old code not updated in US panel.

I used the code I pasted above for timer0 and button_off in my panel and works perfect. It does an if/else in timer0 to send correct command if fan needs to be shut off or just turned on with a percentage. Also removed the prints code at beginning of the button_off because that is redundant and is called when keypress up event enables timer0 (make sure to re-enable timer0 there again for this to work). 

For fun I made buttons bigger using a larger font I found. You should put that font in the HMI. This one is 64h and there is also an 80h.
![image](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/16811687/808c3b94-d037-4286-a9f2-2e169eaf1474)

**Reply by edwardtfn** [reply]
I don't want to remove the timer as it is consistent with the implementation on the climate page (where +/- is also used), so I've made the change on the ESPHome to call `fan.turn_off` when percentage is set to 0%.
It's already available in `dev`. You have to update ESPHome only.

**Reply by edwardtfn** [reply]
I've also fixed the call on button stop on US version. For that you have to update `nspanel_us.tft`.

**Reply by edwardtfn** [reply]
I like the bigger icon, however that will require reorganizing icon distribution on the EU version. I will take a look at this later.

**Reply by Doughboy68** [reply]
Ok I will load and test both dev and your tft and let you know... FYI, the bigger fonts I found in another hmi project and exported.

**Reply by edwardtfn** [reply]
We can use the same font used in other areas with a bigger size, that is not my concern. I'm thinking more about the design and usability using a bigger font and most likely having to move thing around in some pages as we probably want consistence on icon sizes.

**Comment by Doughboy68** [comment]
Tested your new tft and flashed dev and fan works correctly turning off using  all methods on screen.

**Comment by edwardtfn** [comment]
I'm closing this discussion as we have released v4.0 (#1057).
Please add a feature request for the missing things.

And thanks a lot for your support.
